[BEAM-8281] Resize IOITs datasets

mwalenia opened the pull request:
This PR introduces a new way of configuring file-based IOITs. It also adds a known dataset size to reported metrics (the size was measured by the author and is calibrated to be around 1GB) in order to measure also throughput of the IOs (throughput will be calculated on the dashboards)



------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

mwalenia commented:
R: @lgajowy

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/ConfigName.java:
```diff
@@ -0,0 +1,27 @@
+package org.apache.beam.sdk.io.common;
+
+public enum ConfigName {
+  TEXT_1G_GZIP,
```
NIt: could you change 1G to 1GB in this name (does not leave room for speculation)

mwalenia commented:
@lgajowy can you take a look now?

mwalenia commented:
I added smaller test configurations for fast testing after receiving a suggestion on a dev@ thread.
@lgajowy WDYT?

mwalenia commented:
@lgajowy I think your suggestion of allowing users to freely choose their configuration is better than my solution. I modified the classes accordingly and moved the test configurations to Jenkins job definitions. WDYT? After you take a look at the code, please let me know, I'll run the tests to check if everything is fine.

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/avro/AvroIOIT.java:
```diff
@@ -65,7 +64,7 @@
  *  ./gradlew integrationTest -p sdks/java/io/file-based-io-tests
  *  -DintegrationTestPipelineOptions='[
- *  "--numberOfRecords=100000",
+ *  "--testConfigName=AVRO_1GB",
```
This option is missing now. Please update the documentation

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOTestPipelineOptions.java:
```diff
@@ -48,4 +48,15 @@
+  void setExpectedHash(String hash);
+
+  @Description("Size of data saved on the target filesystem")
+  Integer getDatasetSize();
```
I agree that this option should not be required but what if it's null? Will the metric suppliers still save it as null or will the test fail?

lgajowy replied:
What is the unit of size we use in this option (I assume that it's bytes)? This should also be documented.

I propose: `Size of data saved on the target filesystem (bytes)`

lgajowy replied:
Besides that: this should probably eventually land in IOTestPipelineOptions interface but since you are working on file Based tests only in this PR I'm fine with the current state. This is just a note for the future. :)

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/tfrecord/TFRecordIOIT.java:
```diff
@@ -64,9 +63,8 @@
- *  "--numberOfRecords=100000",
- *  "--filenamePrefix=output_file_path",
- *  "--compressionType=GZIP"
+ *  "--testConfigName=TFRECORD_1GB",
```
(similar to above comments)

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/xml/XmlIOIT.java:
```diff
@@ -67,7 +65,7 @@
  *  ./gradlew integrationTest -p sdks/java/io/file-based-io-tests
  *  -DintegrationTestPipelineOptions='[
- *  "--numberOfRecords=100000",
+ *  "--testConfigName=XML_1GB",
```
(as above)

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOTestPipelineOptions.java:
```diff
@@ -48,4 +48,15 @@
+
+  @Validation.Required
+  @Description("Expected hash for write correctness checking")
+  String getExpectedHash();
```
This should probably eventually land in `IOTestPipelineOptions` interface but since you are working on file Based tests only in this PR I'm fine with the current state. This is just a note for the future. :)

lgajowy commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/parquet/ParquetIOIT.java:
```diff
@@ -197,6 +198,9 @@ private void collectAndPublishMetrics(PipelineResult result) {
           return NamedTestResult.create(uuid, timestamp, "run_time", runTime);
         });
 
+    metricSuppliers.add(
```
If `datasetSize` is not provided (this can happen, because the option is not required) this metric value is null. Is this intentional?

mwalenia commented on sdks/java/io/file-based-io-tests/src/test/java/org/apache/beam/sdk/io/common/FileBasedIOTestPipelineOptions.java:
```diff
@@ -48,4 +48,15 @@
   void setReportGcsPerformanceMetrics(boolean performanceMetrics);
+
+  @Validation.Required
+  @Description("Expected hash for write correctness checking")
```
I think your suggestion is better than my description. I'll change it

mwalenia commented:
@lgajowy WDYT now? Should I start running the tests to check if everything works?

mwalenia commented:
Run seed job

mwalenia commented:
Run Java TextIO Performance Test

mwalenia commented:
Run Java TextIO Performance Test
Run Java CompressedTextIO Performance Test
Java ManyFilesTextIO Performance Test
Run Java AvroIO Performance Test
Run Java TFRecordIO Performance Test
Run Java XmlIO Performance Test
Run Java ParquetIO Performance Test
Run Java TextIO Performance Test HDFS
Run Java CompressedTextIO Performance Test HDFS
Run Java ManyFilesTextIO Performance Test HDFS
Run Java AvroIO Performance Test HDFS
Run Java TFRecordIO Performance Test HDFS
Run Java XmlIO Performance Test HDFS
Run Java ParquetIO Performance Test HDFS

mwalenia commented:
Run Java TextIO Performance Test

mwalenia commented:
Run Java CompressedTextIO Performance Test

mwalenia commented:
Run Java ManyFilesTextIO Performance Test

mwalenia commented:
Run Java AvroIO Performance Test

mwalenia commented:
Run Java TFRecordIO Performance Test

mwalenia commented:
Run Java CompressedTextIO Performance Test

mwalenia commented:
Java ManyFilesTextIO Performance Test

mwalenia commented:
Run Java XmlIO Performance Test

mwalenia commented:
Run Java ParquetIO Performance Test

mwalenia commented:
Run Java TextIO Performance Test HDFS

mwalenia commented:
Run Java CompressedTextIO Performance Test HDFS

mwalenia commented:
Run Java ManyFilesTextIO Performance Test HDFS
Run Java AvroIO Performance Test HDFS
Run Java TFRecordIO Performance Test HDFS
Run Java XmlIO Performance Test HDFS
Run Java ParquetIO Performance Test HDFS

mwalenia commented:
Run seed job

mwalenia commented:
Run Java TextIO Performance Test

mwalenia commented:
Run Java CompressedTextIO Performance Test

mwalenia commented:
Run Java ManyFilesTextIO Performance Test

mwalenia commented:
Run Java AvroIO Performance Test

mwalenia commented:
Run Java TFRecordIO Performance Test

mwalenia commented:
Run Java XmlIO Performance Test

mwalenia commented:
Run Java ParquetIO Performance Test

mwalenia commented:
Run Java TextIO Performance Test HDFS

mwalenia commented:
Run Java CompressedTextIO Performance Test HDFS

mwalenia commented:
Run Java ManyFilesTextIO Performance Test HDFS

mwalenia commented:
Run Java AvroIO Performance Test HDFS

mwalenia commented:
Run Java TFRecordIO Performance Test HDFS

mwalenia commented:
Run Java XmlIO Performance Test HDFS
Run Java ParquetIO Performance Test HDFS

mwalenia commented:
Run Java XmlIO Performance Test HDFS

mwalenia commented:
Run Java ParquetIO Performance Test HDFS

mwalenia commented:
Run Java XmlIO Performance Test HDFS

mwalenia commented:
Run seed job

mwalenia commented:
Run Java XmlIO Performance Test HDFS

mwalenia commented:
Run Java ParquetIO Performance Test HDFS

mwalenia commented:
@lgajowy all tests passed, I think we're ready to merge.
